subsys/llext/llext.c

```diff
@@ -0,0 +1,470 @@
+
+		uint32_t stt = ELF_ST_TYPE(sym.st_info);
+		uint32_t stb = ELF_ST_BIND(sym.st_info);
+		uint32_t sect = sym.st_shndx;
+
+		llext_seek(ldr, ldr->sects[LLEXT_SECT_STRTAB].sh_offset + sym.st_name);
+		llext_read(ldr, name, sizeof(name));
+
+		if (stt == STT_FUNC && stb == STB_GLOBAL) {
+			LOG_DBG("function symbol %d, name %s, type tag %d, bind %d, sect %d",
+				i, name, stt, stb, sect);
+			func_syms_cnt++;
+		} else {
+			LOG_DBG("unhandled symbol %d, name %s, type tag %d, bind %d, sect %d",
+				i, name, stt, stb, sect);
+		}
+	}
+
+	/* Copy over global function symbols to symtab */
+
+	ext->sym_tab.syms = k_heap_alloc(&llext_heap, func_syms_cnt * sizeof(struct llext_symbol),
+				       K_NO_WAIT);
+	ext->sym_tab.sym_cnt = func_syms_cnt;
+	pos = ldr->sects[LLEXT_SECT_SYMTAB].sh_offset;
+	j = 0;
+	for (i = 0; i < sym_cnt; i++) {
+		llext_seek(ldr, pos);
+		llext_read(ldr, &sym, ent_size);
+		pos += ent_size;
+
+		uint32_t stt = ELF_ST_TYPE(sym.st_info);
+		uint32_t stb = ELF_ST_BIND(sym.st_info);
+		uint32_t sect = sym.st_shndx;
+
+		llext_seek(ldr, ldr->sects[LLEXT_SECT_STRTAB].sh_offset + sym.st_name);
+		llext_read(ldr, name, sizeof(name));
+
+		if (stt == STT_FUNC && stb == STB_GLOBAL && sect != SHN_UNDEF) {
+			size_t name_sz = sizeof(name);
+
+			ext->sym_tab.syms[j].name = k_heap_alloc(&llext_heap,
+							       sizeof(name),
+							       K_NO_WAIT);
+			strncpy(ext->sym_tab.syms[j].name, name, name_sz);
+			ext->sym_tab.syms[j].addr =
+				(void *)((uintptr_t)ext->mem[ldr->sect_map[sym.st_shndx]]
+					 + sym.st_value);
+			LOG_DBG("function symbol %d name %s addr %p",
+				j, name, ext->sym_tab.syms[j].addr);
+			j++;
+		}
+	}
+
+	/* relocations */
+	uintptr_t loc = 0;
+
+	pos = ldr->hdr.e_shoff;
+
+	for (i = 0; i < ldr->hdr.e_shnum - 1; i++) {
+		llext_seek(ldr, pos);
+		llext_read(ldr, &shdr, sizeof(elf_shdr_t));
+
+		pos += ldr->hdr.e_shentsize;
+
+		/* find relocation sections */
+		if (shdr.sh_type != SHT_REL && shdr.sh_type != SHT_RELA) {
+			continue;
+		}
+
+		rel_cnt = shdr.sh_size / sizeof(elf_rel_t);
+
+
+		llext_seek(ldr, ldr->sects[LLEXT_SECT_SHSTRTAB].sh_offset + shdr.sh_name);
+		llext_read(ldr, name, sizeof(name));
+
+		if (strncmp(name, ".rel.text", sizeof(name)) == 0 ||
+		    strncmp(name, ".rela.text", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_TEXT];
+		} else if (strncmp(name, ".rel.bss", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_BSS];
+		} else if (strncmp(name, ".rel.rodata", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_RODATA];
+		} else if (strncmp(name, ".rel.data", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_DATA];
+		}
+
+		LOG_DBG("relocation section %s (%d) linked to section %d has %d relocations",
+			name, i, shdr.sh_link, rel_cnt);
+
+		for (j = 0; j < rel_cnt; j++) {
+			/* get each relocation entry */
+			llext_seek(ldr, shdr.sh_offset + j * sizeof(elf_rel_t));
+			llext_read(ldr, &rel, sizeof(elf_rel_t));
```

Comment:
should we care about rel type first because link_addr depends of rel type ?
So all relocation calculation depends on arch as relocation type is arch dependant.